enterprise/server/invocation_stat_service/invocation_stat_service.go

```diff
@@ -374,6 +400,264 @@ func (i *InvocationStatService) GetInvocationStatBaseQuery(aggColumn string) str
+	}
+
+	offsetSec := int64(timezoneOffsetMinutes) * 60
+
+	startSec := time.Now().Unix() - ((7 * 24 * time.Hour).Microseconds() / 1e6)
+	if lowTime != nil {
+		startSec = lowTime.GetSeconds()
+	}
+
+	endSec := time.Now().Unix()
+	if highTime != nil {
+		endSec = highTime.GetSeconds()
+	}
+
+	// TODO(jdhollen): Bucket at true midnight in user timezone instead of fudging
+	// it, this gets hosed around daylight savings transitions.  This matters way
+	// less for hourly buckets (where we can easily let the client format).
+	s := time.Unix(startSec, 0).UTC().Add(-time.Minute * time.Duration(timezoneOffsetMinutes))
+	localStart := time.Date(s.Year(), s.Month(), s.Day(), 0, 0, 0, 0, s.Location())
+	currentDateBracket := localStart.Add(time.Minute * time.Duration(timezoneOffsetMinutes))
+
+	currentString := currentDateBracket.Format("2006-01-02")
+	endString := time.Unix(endSec-offsetSec, 0).UTC().AddDate(0, 0, 2).Format("2006-01-02")
+
+	var timestampBuckets []int64
+	for currentString != endString {
+		timestampBuckets = append(timestampBuckets, currentDateBracket.UnixMicro())
+		currentDateBracket = currentDateBracket.AddDate(0, 0, 1)
+		currentString = currentDateBracket.Format("2006-01-02")
+	}
+
+	dStr := make([]string, len(timestampBuckets)-1)
+	for i, d := range timestampBuckets[:len(timestampBuckets)-1] {
+		dStr[i] = fmt.Sprintf("%d", d)
+	}
+	timestampArrayStr := fmt.Sprintf("array(%s)", strings.Join(dStr, ","))
+
+	return timestampBuckets, timestampArrayStr, nil
+}
+
+type HeatmapQueryInputs = struct {
+	PlaceholderBucketQuery string
+	TimestampBuckets       []int64
+	TimestampArrayStr      string
+	MetricBuckets          []int64
+	MetricArrayStr         string
+}
+
+func (i *InvocationStatService) generateQueryInputs(ctx context.Context, table string, metric string, q *stpb.TrendQuery, whereClauseStr string, whereClauseArgs []interface{}, timezoneOffsetMinutes int32) (*HeatmapQueryInputs, error) {
+	timestampBuckets, timestampArrayStr, err := getTimestampBuckets(q, timezoneOffsetMinutes)
+	if err != nil {
+		return nil, err
+	}
+	metricBuckets, metricArrayStr, err := i.getMetricBuckets(ctx, table, metric, whereClauseStr, whereClauseArgs)
+	if err != nil {
+		return nil, err
+	}
+	if len(metricBuckets) == 0 {
+		return nil, nil
+	}
+
+	pbq := fmt.Sprintf(`
+	  SELECT toInt64(timestamp) as timestamp, toInt64(bucket) as bucket, 0 as v FROM (
+			SELECT arrayElement(%s, number + 1) AS timestamp FROM numbers(%d)) AS a
+		CROSS JOIN (
+			SELECT arrayElement(%s, number + 1) AS bucket FROM numbers(%d)) AS b
+		ORDER BY timestamp, bucket`,
+		timestampArrayStr, len(timestampBuckets)-1, metricArrayStr, len(metricBuckets)-1)
+
+	return &HeatmapQueryInputs{
+		PlaceholderBucketQuery: pbq,
+		TimestampBuckets:       timestampBuckets,
+		TimestampArrayStr:      timestampArrayStr,
+		MetricBuckets:          metricBuckets,
+		MetricArrayStr:         metricArrayStr}, nil
```

Comment:
nit: should "}, nil" go to the next line?

Reply:
This is an autoformatter quirk of mine, I did something different now, I have basically no opinion on this :P